src/ty.md

```diff
@@ -63,6 +68,67 @@ defining all of the different kinds of types in the compiler.
+
+**Slide**
+Corresponds to `[T]`.
+
+**Array**
+Corresponds to `[T; n]`.
+
+**RawPtr**
+Corresponds to `*mut T` or `*const T`
+
+**Ref**
+`Ref` stands for safe references, `&'a mut T` or `&'a T`. `Ref` has some associated parts, like `Ty<'tcx>` which is the type that the reference references, `Region<'tcx>` is the lifetime or region of the reference and `Mutability` if the reference is mutable or not.
+
+
+## Interning
+
+We create a LOT of types during compilation. For performance reasons, we allocate them from a global memory pool, they are each allocated once from a long-lived *arena*. This system reduces allocations/deallocations of memory. It also allows for easy comparison of types for equality: we implemented `PartialEq for TyS` ([link](https://github.com/rust-lang/rust/blob/3ee936378662bd2e74be951d6a7011a95a6bd84d/src/librustc/ty/mod.rs#L528-L534)), so we can just compare pointers. The `CtxtInterners` ([link](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.CtxtInterners.html#structfield.arena)) type contains a bunch of maps of interned types and the arena itself.
+
+Each time we want to construct a type, the compiler doesn’t naively allocate from the buffer but we look if that type was already constructed. If it was, we just get the same pointer we had before, otherwise we make a fresh pointer.
+With this schema if we want to know if two types are the same, all we need to do is compare the pointers which is efficient. `TyS` which represents types is carefully setup so you never construct them on the stack. You always allocate them from this arena and you always intern them so they are unique. This also means that `PartialEq` is implemented by just comparing the pointers of the two of them.
+
+At the beginning of the compilation we make a buffer and each time we need to allocate a type we use some of this memory buffer and if we run out of space we get another one. The lifetime of that buffer is `'tcx`. Our types are tied to that lifetime, so when compilation finishes all the memory related to that buffer is freed and our `'tcx` references would be invalid. This is called arena allocation.
+
+However, there are a lot of types that we don’t want to keep around for the entire compilation. For example, when we are type-checking a specific function,  we will have a bunch of local types (for things in the body of the function) that are irrelevant outside the function. We don’t need them around after we finish type checking the function, and keeping them around would just waste memory. For that reason, we actually have two types of arenas: the *global arena* is for types that are relevant everywhere (e.g. `u32` or a `struct` definition) and the fresh *local arena* whenever we start to type check specific function. Local types of the function are allocated in the local arena,  and when we are done, we can discard the entire arena, freeing up the space.
```

Comment:
There's not a local arena any more.

Reply:
@matthewjasper ideas on how can we adapt the text?.

Reply:
I think that this paragraph can be removed.

Reply:
Done!